beets/ui/__init__.py

```diff
@@ -1180,55 +863,108 @@ def _open_library(config):
-                      help=u'log more details (use twice for even more)')
-    parser.add_option('-c', '--config', dest='config',
-                      help=u'path to configuration file')
-    parser.add_option('-h', '--help', dest='help', action='store_true',
-                      help=u'show this help message and exit')
-    parser.add_option('--version', dest='version', action='store_true',
-                      help=optparse.SUPPRESS_HELP)
-
-    options, subargs = parser.parse_global_options(args)
-
-    # Special case for the `config --edit` command: bypass _setup so
-    # that an invalid configuration does not prevent the editor from
-    # starting.
-    if subargs and subargs[0] == 'config' \
-       and ('-e' in subargs or '--edit' in subargs):
-        from beets.ui.commands import config_edit
-        return config_edit()
-
-    test_lib = bool(lib)
-    subcommands, plugins, lib = _setup(options, lib)
-    parser.add_subcommand(*subcommands)
-
-    subcommand, suboptions, subargs = parser.parse_subcommand(subargs)
-    subcommand.func(lib, suboptions, subargs)
-
-    plugins.send('cli_exit', lib=lib)
-    if not test_lib:
-        # Clean up the library unless it came from the test harness.
-        lib._close()
-
-
-def main(args=None):
-    """Run the main command-line interface for beets. Includes top-level
-    exception handlers that print friendly error messages.
+    def load_commands(self):
+        """Using the current beets configuration, load the set of
+        subcommands for the `beet` executable. Any previously loaded
+        commands in `self.commands` are cleared.
+        """
+        # The built-in commands.
+        from beets.ui.commands import default_commands
+        subcommands = list(default_commands)
+
+        # Commands from plugins.
+        subcommands.extend(plugins.commands())
+
+        # Populate the commands dict.
+        self.commands = {}
+        for command in subcommands:
+            # For compatibility, convert old optparse-based commands
+            # into Click commands.
+            if isinstance(command, cli.LegacySubcommand):
+                command = command._to_click()
+            self.add_command(command)
+
+    def resolve_command(self, ctx, args):
+        # Special case for the `config --edit` command. This bypasses
+        # all of the normal beets initialization so the command can
+        # succeed even with a broken configuration.
+        str_args = [click.utils.make_str(a) for a in args]
+        if str_args[0] == 'config' and ('-e' in str_args or
+                                        '--edit' in str_args):
+            from beets.ui.commands import config_cmd
+            command = config_cmd._to_click()
+            return str_args[0], command, args[1:]
+
+        # In the normal case, configure beets and then load the set of
+        # subcommands immediately before continuing with the resolution
+        # logic.
+        config = _configure(ctx.params)
```

Comment:
ths is is eating the verbose configuration from the logging tests (and I'm sure other places). 

Click sets verbose to 0 since we have no default and then writes right over the parameter.

```python
    def test_command_level1(self):
        self.config['verbose'] = 1
        with helper.capture_log() as logs:
            self.run_command('dummy')
```
Do we want to use dynamic defaults for these parameters directly in `click.option`? or solve it at a higher level? I still don't get click well enough to know what would be best.

Our current configuration mechanism is kind of spooky action at a distance.